Optionally handle subclasses of exceptions in custom error handler

nickdnk opened the pull request:
Hello guys

I thought this could be a nice addition, so a particular app can decide to handle all subclasses of an exception with one handler. This is nice way to keep things separated.

I added this as an option that defaults to false so this *should* not be a breaking change, but I'm not 100% sure since `$handlers` is `protected` and not `private`.

What do you think?

nickdnk commented:
Force pushed for typo.

nickdnk commented:
We could mitigate the BC by doing this instead:

```
if (isset($this->handlers[$type])) {
    return $this->callableResolver->resolve(
        is_array($this->handlers[$type]) ? $this->handlers[$type]['handler'] : $this->handlers[$type]
    );
} else {
    foreach ($this->handlers as $class => $handler) {
        if (is_array($handler) && $handler['handle_subclass'] === true) {
            if (is_subclass_of($type, $class)) {
                return $this->callableResolver->resolve($handler['handler']);
            }
        }
    }
}
```

Then even if a someone has subclassed ErrorMiddleware *and* overriden $handlers to contain callables, it wouldn't break.

l0gicgate commented:
@adriansuter I think this would be a good addition and there's no BC breaks. What do you think about it?

nickdnk commented:
Is there a reason that `$handlers` is `protected` and not `private` ? Given how close this variable is tied to the `setErrorHandler` method, should it be visible to a subclass?

l0gicgate commented:
> Is there a reason that $handlers is protected and not private ? Given how close this variable is tied to the setErrorHandler method, should it be visible to a subclass?

In case one wants to extend `ErrorMiddleware` we left it as protected.

nickdnk commented:
I think yours is a good workaround, @adriansuter.

l0gicgate commented:
@adriansuter I like your solution. We should proceed with that.

nickdnk commented:
Updated the PR.

Also, I added phpdocs for the handler arrays based on the value of `$defaultErrorHandler` - but I'm not entirely sure that you can annotate them as `ErrorHandlerInterface` since they're functions?

l0gicgate commented:
Need to write tests as well for the subclass case. Are you able to do that?

nickdnk commented:
I did. I don't know why it claims it wasn't covered. There is a test for both with and without subclass use.

adriansuter commented:
Did you add a test case for the case you set an error handler as subclass handler, but the exception thrown is not a subclass but exactly that class?

`} elseif (isset($this->subClassHandlers[$type])) {`

nickdnk commented:
I think that did it.

l0gicgate commented on tests/Middleware/ErrorMiddlewareTest.php:
```diff
@@ -99,24 +99,129 @@ public function testGetErrorHandlerWillReturnDefaultErrorHandlerForUnhandledExce
+
+    public function testSuperclassExceptionHandlerHandlesSubclassException()
+    {
+
```
Can you remove this blank line please.

l0gicgate commented on tests/Middleware/ErrorMiddlewareTest.php:
```diff
@@ -99,24 +99,129 @@ public function testGetErrorHandlerWillReturnDefaultErrorHandlerForUnhandledExce
+
+    public function testSuperclassExceptionHandlerDoesNotHandleSubclassException()
+    {
+
```
Can you remove this blank line please.
